### PR TITLE
Adds Traffic Manager Terraform and pipeline.

### PR DESCRIPTION
**Pull request by TomKimber**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/DTSPO-3357

### Change description ###

Adds Traffic Manager Terraform and pipeline. Should do nothing without config.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by luigibk — APPROVED** [review]

File: azure-pipelines-trafficmanager.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
+name: SDS Azure Platform - Traffic Manager
```
Comment (timja) [inline azure-pipelines-trafficmanager.yaml:1]: why do you need your own pipeline?

if you do need it then you need to fiddle with files_list so the right PRs trigger on each PR but i wouldn't have thought you would need it

File: azure-pipelines-trafficmanager.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    displayName: Location
+    type: string
+    default: "UK South"
+    values:
+      - "UK South"
+      - "UK West"
+
+variables:
+  - name: tfversion
+    value: 0.15.1
```
Comment (timja) [inline azure-pipelines-trafficmanager.yaml:42]:
```suggestion
    value: 1.0.0
```

File: azure-pipelines-trafficmanager.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  - name: project
+    value: application
+  - name: timeoutInMinutes
+    value: "360"
+  - name: agentPool
+    value: "hmcts-cftptl-agent-pool"
```
Comment (timja) [inline azure-pipelines-trafficmanager.yaml:48]: do you need the private agent?

File: components/trafficmanager/init.tf [hunk 4]
```diff
@@ -0,0 +1,19 @@
+terraform {
+  required_version = ">= 0.15.1"
+
+  backend "azurerm" {
+    subscription_id = "04d27a32-7a07-48b3-95b8-3c8691e1a263"
+  }
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "2.58.0"
```
Comment (timja) [inline components/trafficmanager/init.tf:10]: use latest?

File: components/trafficmanager/main.tf [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+module "traffic_manager" {
+  providers = {
+    azurerm      = azurerm
+    azurerm.data = azurerm.data
+  }
+  source = "git::https://github.com/hmcts/terraform-module-frontdoor.git?ref=master"
+
+  # General
```
Comment (timja) [inline components/trafficmanager/main.tf:15]: remove useless comments
```suggestion
```

File: components/trafficmanager/main.tf [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  common_tags = module.ctags.common_tags
+
+  # Traffic Manager Endpoints
```
Comment (timja) [inline components/trafficmanager/main.tf:18]: remove useless comments
```suggestion
```

File: components/trafficmanager/main.tf [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  traffic_manager_endpoints = var.traffic_manager_endpoints
+
+  # Traffic Manager Profiles
```
Comment (timja) [inline components/trafficmanager/main.tf:21]:
```suggestion
  # Traffic Manager Profiles
```

File: environments/variable.tf [hunk 8]
```diff
@@ -15,3 +15,5 @@ variable "vnet_name" {}
 variable "vnet_rg" {}
 variable "product" {}
 variable "builtFrom" {}
+variable "traffic_manager_endpoints" {}
```
Comment (timja) [inline environments/variable.tf:18]: won't this fail because you haven't defined any?

File: components/trafficmanager/init.tf (hunk 4 above)
Comment (TomKimber) [reply components/trafficmanager/init.tf]: Bumped Azure provider to 2.64.0.

File: environments/variable.tf (hunk 8 above)
Comment (TomKimber) [reply environments/variable.tf:18]: It shouldn't fail - It should do nothing. I'll add the configuration shortly but I need the pipeline first so figured I'd add the Terraform at the same time.

File: azure-pipelines-trafficmanager.yaml (hunk 3 above)
Comment (TomKimber) [reply azure-pipelines-trafficmanager.yaml:48]: No, I can use Microsoft hosted agents.

File: azure-pipelines-trafficmanager.yaml (hunk 1 above)
Comment (TomKimber) [reply azure-pipelines-trafficmanager.yaml]: Removed Traffic Manage pipeline and added to base pipeline.

File: azure-pipelines.yaml [hunk 9]
```diff
@@ -10,66 +10,69 @@ parameters:
...
-      - deployment: 'appgateway'
-        environment: 'ithc'
-        dependsOn: 'Precheck'
-      - deployment: 'appgateway'
-        environment: 'test'
-        dependsOn: 'Precheck'
-      - deployment: 'appgateway'
-        environment: 'prod'
-        dependsOn: 'Precheck'
+      - deployment: "backendappgateway"
```
Comment (timja) [inline azure-pipelines.yaml:13]: did you run some auto-formatter over this? I can't see any reason for any changes in this file?

File: environments/variable.tf [hunk 10]
```diff
@@ -15,3 +15,5 @@ variable "vnet_name" {}
 variable "vnet_rg" {}
 variable "product" {}
 variable "builtFrom" {}
+variable "traffic_manager_endpoints" { default = {} }
```
Comment (timja) [inline environments/variable.tf:18]: is this formatted properly?

File: azure-pipelines.yaml (hunk 9 above)
Comment (TomKimber) [reply azure-pipelines.yaml:13]: Yes, it's linting on save. I can revert it if you'd prefer?

File: environments/variable.tf (hunk 10 above)
Comment (TomKimber) [reply environments/variable.tf:18]: Yes, I believe so. The module is now working as expected.

File: azure-pipelines.yaml (hunk 9 above)
Comment (TomKimber) [reply azure-pipelines.yaml]: Reverted linting.

**Review by timja — APPROVED** [review]
i would still recommend removing those traffic manager comments, they literally repeat the variable name

**Comment by TomKimber** [comment]
In the current module definition - they repeat the variable name but they're included to indicate where to add the additional blocks if required. For example:

`  # General
  common_tags         = module.ctags.common_tags
  resource_group_name = var.resource_group_name

  # Traffic Manager Endpoints
  traffic_manager_endpoints               = var.traffic_manager_endpoints
  traffic_manager_endpoint_custom_headers = var.traffic_manager_endpoint_custom_headers
  traffic_manager_endpoint_subnets        = var.traffic_manager_endpoint_subnets

  # Traffic Manager Profiles
  traffic_manager_profiles               = var.traffic_manager_profiles
  traffic_manager_profile_custom_headers = var.traffic_manager_profile_custom_headers`

**Comment by timja** [comment]
> In the current module definition - they repeat the variable name but they're included to indicate where to add the additional blocks if required. For example:
> 
> ```
>   # General
>   common_tags         = module.ctags.common_tags
>   resource_group_name = var.resource_group_name
> 
>   # Traffic Manager Endpoints
>   traffic_manager_endpoints               = var.traffic_manager_endpoints
>   traffic_manager_endpoint_custom_headers = var.traffic_manager_endpoint_custom_headers
>   traffic_manager_endpoint_subnets        = var.traffic_manager_endpoint_subnets
> 
>   # Traffic Manager Profiles
>   traffic_manager_profiles               = var.traffic_manager_profiles
>   traffic_manager_profile_custom_headers = var.traffic_manager_profile_custom_headers
> ```
> 
> I think it makes the structure neater and more readable but I'm not wedded to them.

Add it when you need it 👅 but not blocking on that

**Pull request merged**